packages/fetchai/protocols/fipa/dialogues.py

```diff
@@ -60,53 +59,48 @@
     ],
 }  # type: Dict[FipaMessage.Performative, List[Union[None, FipaMessage.Performative]]]
 
+SUPPLY_DATAMODEL_NAME = "supply"
+DEMAND_DATAMODEL_NAME = "demand"
+
 
 class FipaDialogue(Dialogue):
     """The FIPA dialogue class maintains state of a dialogue and manages it."""
 
-    STARTING_MESSAGE_ID = 1
-    STARTING_TARGET = 0
-
-    class EndState(Enum):
+    class EndState(Dialogue.EndState):
         """This class defines the end states of a dialogue."""
 
         SUCCESSFUL = 0
         DECLINED_CFP = 1
         DECLINED_PROPOSE = 2
         DECLINED_ACCEPT = 3
 
-    class AgentRole(Enum):
-        """This class defines the agent's role in the dialogue."""
+    class AgentRole(Dialogue.Role):
+        """This class defines the agent's role in a fipa dialogue."""
 
         SELLER = "seller"
         BUYER = "buyer"
 
     def __init__(
-        self, dialogue_label: DialogueLabel, is_seller: bool, **kwargs
+        self, dialogue_label: DialogueLabel, role: AgentRole, **kwargs
     ) -> None:
         """
-        Initialize a dialogue label.
+        Initialize a fipa dialogue.
 
         :param dialogue_label: the identifier of the dialogue
-        :param is_seller: indicates whether the agent associated with the dialogue is a seller or buyer
+        :param role: the role of the agent this dialogue is maintained for
 
         :return: None
         """
-        Dialogue.__init__(self, dialogue_label=dialogue_label)
-        self._is_seller = is_seller
-        self._role = (
-            FipaDialogue.AgentRole.SELLER if is_seller else FipaDialogue.AgentRole.BUYER
-        )
+        Dialogue.__init__(self, dialogue_label=dialogue_label, role=role)
 
-    @property
-    def is_seller(self) -> bool:
-        """Check whether the agent acts as the seller in this dialogue."""
-        return self._is_seller
+    @staticmethod
+    def role_from_first_message(message: Message) -> Dialogue.Role:
+        """Infer the role of the agent from an incoming/outgoing first message
 
-    @property
-    def role(self) -> "FipaDialogue.AgentRole":
-        """Get role of agent in dialogue."""
-        return self._role
+        :param message: an incoming/outgoing first message
+        :return: The role of the agent
+        """
+        pass
```

Comment:
```suggestion
        role = role_from_first_message(message)
        self.role = role
```
Then `role_from_first_message` can be imported from `custom_types.py`

Reply:
Done